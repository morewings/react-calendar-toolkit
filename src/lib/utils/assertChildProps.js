import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
