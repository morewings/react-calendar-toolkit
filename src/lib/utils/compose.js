const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
