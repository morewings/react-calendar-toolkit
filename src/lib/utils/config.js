export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
