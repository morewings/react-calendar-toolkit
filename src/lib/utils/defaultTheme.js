const baseColor = '#43a047';
const gray = 'rgba(0, 0, 0, 0.38)';
const borderRadius = '4px';

const defaultTheme = {
  /** Set font for datepicker */
  fontPrimary: 'Helvetica Neue',
  /** Font, which is applied, if --fontPrimary is not available */
  fontSecondary: 'Arial',
  /** Fallback font if two previous are unavailable */
  fontFallback: 'sans-serif',
  fontSize: '14px',
  fontWeight: '400',
  /** Applied to the title */
  fontSizeSmall: '10px',
  /** Applied to the header text */
  fontSizeBig: '24px',
  /** Set inner width */
  calendarWidth: '240px',
  /** Calculate total width */
  datePickerWidth: 'calc(var(--calendarWidth) + var(--innerPadding) * 2)',
  selectorHeight: '36px',
  calendarHeight: '200px',
  /** Set padding between elements */
  innerPadding: '12px',
  /** Day, month, year entries height */
  entryHeight: '32px',
  entryMargin: '1px',
  datepickerBorderRadius: borderRadius,
  entryBorderRadius: borderRadius,
  /** Background color of Calendar */
  bgColor: 'white',
  /** Base color for text */
  textColor: 'black',
  headerBgColor: baseColor,
  weekendTextColor: 'red',
  headerTextColor: 'white',
  borderColor: gray,
  modalBgColor: gray,
  weekDayColor: gray,
  entryHoverColor: 'rgba(0, 0, 0, 0.08)',
  entrySelectedColor: baseColor,
  entryHighlightColor: 'lightblue',
  inputWidth: '90px',
  inputPadding: '8px',
  inputBorderRadius: borderRadius,
  inputBorderColor: gray,
  inputHighlightColor: 'lightblue',
};

export default defaultTheme;
