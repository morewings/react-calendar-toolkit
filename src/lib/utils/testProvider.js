import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
