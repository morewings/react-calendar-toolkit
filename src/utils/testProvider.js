import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'features/theme/withTheme';
import {Provider as ModalProvider} from 'features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
