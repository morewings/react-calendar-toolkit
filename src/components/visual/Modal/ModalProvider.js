import React, {useEffect, useRef, useState, Fragment} from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import {setCSSVariable, useThemeContext} from 'utils/themeContext';
import defaults from 'utils/defaultTheme';
import classes from './Modal.module.css';

const Modal = ({
  toggleDatepicker,
  children,
  isVisible,
  renderDatePickerAs,
  wrapModalWith,
}) => {
  const modalContainerRef = useRef();
  const [isNodeAttached, setIsNodeAttached] = useState(false);
  const Datepicker = renderDatePickerAs;
  const Wrapper = wrapModalWith;
  const theme = useThemeContext();

  /** Create modal container node */
  useEffect(() => {
    const propName = '--modalBgColor';
    const bgColor = theme[propName] || defaults[propName];
    if (isVisible && !isNodeAttached) {
      modalContainerRef.current = document.createElement('div');
      modalContainerRef.current.classList.add(classes.container);
      setCSSVariable(modalContainerRef.current, propName, bgColor);
      document.body.appendChild(modalContainerRef.current);
      setIsNodeAttached(true);
    }
    return () => {
      isNodeAttached && modalContainerRef.current.remove();
      isNodeAttached && setIsNodeAttached(false);
    };
  }, [isNodeAttached, isVisible, theme]);

  return (
    <Fragment>
      {children}
      {isVisible &&
        isNodeAttached &&
        ReactDOM.createPortal(
          <Wrapper toggleDatepicker={toggleDatepicker}>
            <Datepicker />
          </Wrapper>,
          modalContainerRef.current
        )}
    </Fragment>
  );
};

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  toggleDatepicker: PropTypes.func.isRequired,
  renderDatePickerAs: PropTypes.elementType.isRequired,
  wrapModalWith: PropTypes.elementType.isRequired,
  isVisible: PropTypes.bool.isRequired,
};

export default Modal;
