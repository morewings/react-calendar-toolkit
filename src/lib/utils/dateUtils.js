import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
