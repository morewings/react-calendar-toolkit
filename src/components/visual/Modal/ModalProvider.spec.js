import React from 'react';
import {InputMockProvider as Provider} from 'utils/testProvider';
import {render} from '@testing-library/react';
import ModalProvider from './ModalProvider';

const Datepicker = () => <div data-testid="datepicker">Datepicker</div>;
const Input = () => <div data-testid="input">Input</div>;
const Wrapper = ({children}) => <div data-testid="wrapper">{children}</div>;

const toggleDatepicker = jest.fn();

const defaultProps = {
  toggleDatepicker,
  isVisible: false,
  renderDatePickerAs: Datepicker,
  wrapModalWith: Wrapper,
};

describe('ModalProvider', () => {
  it('renders without modal', () => {
    render(
      <ModalProvider {...defaultProps}>
        <Input />
      </ModalProvider>,
      {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    expect(document.body).toMatchSnapshot();
  });

  it('renders with modal', () => {
    const props = {
      ...defaultProps,
      isVisible: true,
    };
    render(
      <ModalProvider {...props}>
        <Input />
      </ModalProvider>,
      {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    expect(document.body).toMatchSnapshot();
  });
});
