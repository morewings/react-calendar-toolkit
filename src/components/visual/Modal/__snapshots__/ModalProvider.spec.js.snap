// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider renders with modal 1`] = `
<body>
  <div>
    <div
      data-testid="input"
    >
      Input
    </div>
  </div>
  <div
    class="container"
    style="--modalBgColor: rgba(0, 0, 0, 0.38);"
  >
    <div
      data-testid="wrapper"
    >
      <div
        data-testid="datepicker"
      >
        Datepicker
      </div>
    </div>
  </div>
</body>
`;

exports[`ModalProvider renders without modal 1`] = `
<body>
  <div>
    <div
      data-testid="input"
    >
      Input
    </div>
  </div>
</body>
`;
